src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs

```diff
@@ -0,0 +1,1340 @@
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            return Interop.Secur32.SspiFreeAuthIdentity(handle) == Interop.SecurityStatus.OK;
+        }
+    }
+
+    //
+    //  A set of Safe Handles that depend on native FreeContextBuffer finalizer
+    //
+#if DEBUG
+    internal abstract class SafeFreeContextBuffer : DebugSafeHandle
+    {
+#else
+    internal abstract class SafeFreeContextBuffer : SafeHandleZeroOrMinusOneIsInvalid
+    {
+#endif
+        protected SafeFreeContextBuffer() : base(true) { }
+
+        // This must be ONLY called from this file and in the context of a CER
+        internal unsafe void Set(IntPtr value)
+        {
+            this.handle = value;
+        }
+
+        internal static int EnumeratePackages(out int pkgnum, out SafeFreeContextBuffer pkgArray)
+        {
+            int res = -1;
+            SafeFreeContextBuffer_SECURITY pkgArray_SECURITY = null;
+            res = Interop.Secur32.EnumerateSecurityPackagesW(out pkgnum, out pkgArray_SECURITY);
+            pkgArray = pkgArray_SECURITY;
+
+            if (res != 0 && pkgArray != null)
+            {
+                pkgArray.SetHandleAsInvalid();
+            }
+
+            return res;
+        }
+
+        internal static SafeFreeContextBuffer CreateEmptyHandle()
+        {
+            return new SafeFreeContextBuffer_SECURITY();
+        }
+
+        //
+        // After PInvoke call the method will fix the refHandle.handle with the returned value.
+        // The caller is responsible for creating a correct SafeHandle template or null can be passed if no handle is returned.
+        //
+        // This method switches between three non-interruptible helper methods.  (This method can't be both non-interruptible and
+        // reference imports from all three DLLs - doing so would cause all three DLLs to try to be bound to.)
+        //
+        public unsafe static int QueryContextAttributes(SafeDeleteContext phContext, Interop.Secur32.ContextAttribute contextAttribute, byte* buffer, SafeHandle refHandle)
+        {
+            return QueryContextAttributes_SECURITY(phContext, contextAttribute, buffer, refHandle);
+        }
+
+        private unsafe static int QueryContextAttributes_SECURITY(
+            SafeDeleteContext phContext,
+            Interop.Secur32.ContextAttribute contextAttribute,
+            byte* buffer,
+            SafeHandle refHandle)
+        {
+            int status = (int)Interop.SecurityStatus.InvalidHandle;
+            bool b = false;
+
+            // We don't want to be interrupted by thread abort exceptions or unexpected out-of-memory errors failing to jit
+            // one of the following methods. So run within a CER non-interruptible block.
+            try
+            {
+                phContext.DangerousAddRef(ref b);
+            }
+            catch (Exception e)
+            {
+                if (b)
+                {
+                    phContext.DangerousRelease();
```

Comment:
Good catch: I'll scan for all other places where AddRef/Release are used.

In a previous discussion with @davidsh we've found that since in CoreFX ThreadAbort can't be caught, the entire pattern can be removed.